### PR TITLE
Feature: Game logic

### PR DESCRIPTION
**Pull request by moinkhanif**
In this feature branch, we created and executed the game logic behind the tic-tac-toe game. We created the following files:

- board.rb
- game.rb
- interface.rb
- player.rb

### CODE REVIEW

**Comment by Anna-Myzukina** [comment]
# Status: [changes required]


### Good job with the milestone. Few steps and your project can be approved:


* [x] No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

 * [x] Using require_relative for files in your directory instead of require is a ruby convention.

`"Require should be used for external files, like gems, while require_relative should be used for referring to files within your directory."`

* [x] Use attr_readers/attr_writers instead of attr_accessors. The student must be aware that by using it they allow writing that property from outside the class. That variable can be set to anything. <TODO: explain why> ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/attr_accessor.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))


### Make changes and submit another code review request.

*If you have any questions, let me know on slack (@Anna-Muzykina) and I will be glad to clarify with you.*
*Happy coding * <g-emoji class="g-emoji" alias="computer" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4bb.png">computer</g-emoji>

**Comment by moinkhanif** [comment]
> # Status: [changes required]
> ### Good job with the milestone. Few steps and your project can be approved:
> * [x]  No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
> * [x]  Using require_relative for files in your directory instead of require is a ruby convention.
> 
> `"Require should be used for external files, like gems, while require_relative should be used for referring to files within your directory."`
> 
> * [x]  Use attr_readers/attr_writers instead of attr_accessors. The student must be aware that by using it they allow writing that property from outside the class. That variable can be set to anything. <TODO: explain why> ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/attr_accessor.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
> 
> ### Make changes and submit another code review request.
> _If you have any questions, let me know on slack (@Anna-Muzykina) and I will be glad to clarify with you._
> *Happy coding * computer

Thank you for pointing it out. I have corrected my mistakes 💯

File: bin/main.rb [hunk 1]
```diff
@@ -1,19 +1,20 @@
...
+    puts
     puts "\t  #{positions[6]} | #{positions[7]} | #{positions[8]}"
     puts "\t ---|---|---"
     puts "\t  #{positions[3]} | #{positions[4]} | #{positions[5]}"
     puts "\t ---|---|---"
-    puts "\t  #{positions[0]} | #{positions[1]} | #{positions[2]}"
-    puts ''
+    puts "\t  #{positions[0]} | #{positions[1]} | #{positions[2]}"
+    puts
   end
```
Comment (euqueme) [inline bin/main.rb:18]: [OPTIONAL]

- [ ] DRY rule should be respected. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/dry.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- I think the code in your show method is repetitive and  could be done with a loop of some sort, please consider changing it.

**Review by euqueme — APPROVED** [review]
> ### Pul Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @moinkhanif and @Samkiroko :balloon: :lollipop: 

### Thank you for doing the changes requested by the previous reviewer :medal_sports: your game plays great and I believe you have implemented good coding practices so there's no need to keep you here any longer so I am approving your project :heavy_check_mark: 

## OPTIONAL

#### Behaviour

- [ ] The game lets you have empty names or the same name for both players, let's add validations or set default names when this happens

image

#### Implementation

- [ ] I see you have module Interface inside your main.rb file, and even though in the project specs it doesn't say anything about where to place modules I think it should be kept in a separate file.

* [ ] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing). ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/single_responsibility.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

- the next_move method within your interface module has several responsibilities, I strongly recommend to break it down in multiple methods or put the code directly in the game loop within your main.rb file  (remove this method from your interface module) to respect this rule 

```ruby
 def next_move(player)
    loop do
      print "#{player.name}, "
      input = get_input('position')

      if valid_number?(input)
        input = input.to_i
        if @board.taken?(input)
          get_input('taken')
        else
          player.move(input)
          @board.update(input, player.char)
          show(@board.positions)
          break
        end
      else
        get_input('quit') if input == 'q'
        get_input('invalid')
        sleep 1
      end
      show(@board.positions)
    end
  end
```
#### General

- [x] For future projects, make sure both students add even amount of commits, remember these projects are going to be linked in your portfolio and future employers want to see proof of what we are capable of doing :muscle: 

### That's it, team! :raised_hands: 

### Please :pray: do the changes (or not) :man_technologist: :man_technologist: and merge this branch when you are done :memo: 

Happy coding :rocket: 

Any questions? is better to ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: APPROVED :green_apple:

**Pull request merged**